src/machine_learning/loss_function/negative_log_likelihood.rs

```diff
@@ -0,0 +1,128 @@
+        let loss: f64 = -a * p.ln() - (1.0 - a) * (1.0 - p).ln();
+        total_loss += loss;
+    }
+    Ok(total_loss / (y_pred.len() as f64))
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum NegativeLogLikelihoodLossError {
+    InputsHaveDifferentLength,
+    EmptyInputs,
+    InvalidValues,
+}
+
+fn are_all_values_in_range(values: &[f64]) -> bool {
+    values.iter().all(|&x| (0.0..=1.0).contains(&x))
+}
+
+#[cfg(test)]
+mod tests {
+
+    use super::*;
+
+    #[test]
+    fn test_neg_log_likelihood() {
+        let actual_values: Vec<f64> = vec![1.0, 0.0, 1.0];
+        let predicted_values: Vec<f64> = vec![0.9, 0.1, 0.8];
+        assert_eq!(
+            neg_log_likelihood(&actual_values, &predicted_values),
+            Ok(0.14462152754328741)
+        );
+    }
+
+    #[test]
+    fn test_neg_log_likelihood_invalid_length0() {
+        let actual_values: Vec<f64> = vec![1.0, 0.0, 1.0];
+        let predicted_values: Vec<f64> = vec![0.9, 0.1];
+        assert_eq!(
+            neg_log_likelihood(&actual_values, &predicted_values),
+            Err(NegativeLogLikelihoodLossError::InputsHaveDifferentLength)
+        );
+    }
+
+    #[test]
+    fn test_neg_log_likelihood_invalid_length1() {
+        let actual_values: Vec<f64> = vec![1.0, 0.0];
+        let predicted_values: Vec<f64> = vec![0.9, 0.1, 0.8];
+        assert_eq!(
+            neg_log_likelihood(&actual_values, &predicted_values),
+            Err(NegativeLogLikelihoodLossError::InputsHaveDifferentLength)
+        );
+    }
+
+    #[test]
+    fn test_neg_log_likelihood_invalid_values() {
+        let actual_values: Vec<f64> = vec![1.0, 0.0, 1.0];
+        let predicted_values: Vec<f64> = vec![1.1, 0.1, 0.8];
+        assert_eq!(
+            neg_log_likelihood(&actual_values, &predicted_values),
+            Err(NegativeLogLikelihoodLossError::InvalidValues)
+        );
+    }
+
+    #[test]
+    fn test_neg_log_likelihood_empty_prediction() {
+        let actual_values: Vec<f64> = vec![1.0, 0.0, 1.0];
+        let predicted_values: Vec<f64> = vec![];
+        assert_eq!(
+            neg_log_likelihood(&actual_values, &predicted_values),
+            Err(NegativeLogLikelihoodLossError::EmptyInputs)
+        );
+    }
+
+    #[test]
+    fn test_neg_log_likelihood_negative_values0() {
+        let actual_values: Vec<f64> = vec![-1.0, 0.0, 1.0];
+        let predicted_values: Vec<f64> = vec![0.9, 0.1, 0.8];
+        assert_eq!(
+            neg_log_likelihood(&actual_values, &predicted_values),
+            Err(NegativeLogLikelihoodLossError::InvalidValues)
+        );
+    }
+
+    #[test]
+    fn test_neg_log_likelihood_negative_values1() {
+        let actual_values: Vec<f64> = vec![1.0, 0.0, 1.0];
+        let predicted_values: Vec<f64> = vec![0.9, 0.1, -0.8];
+        assert_eq!(
+            neg_log_likelihood(&actual_values, &predicted_values),
+            Err(NegativeLogLikelihoodLossError::InvalidValues)
+        );
+    }
```

Comment:
I would try again:

```suggestion
    macro_rules! test_with_wrong_inputs {
        ($($name:ident: $inputs:expr,)*) => {
            $(
                #[test]
                fn $name() {
                    let (values_a, values_b, expected_error) = $inputs;
                    assert_eq!(neg_log_likelihood(&values_a, &values_b), expected_error);
                    assert_eq!(neg_log_likelihood(&values_b, &values_a), expected_error);
                }
            )*
        }
    }

    test_with_wrong_inputs! {
        different_length: (vec![0.9, 0.0, 0.8], vec![0.9, 0.1], Err(NegativeLogLikelihoodLossError::InputsHaveDifferentLength)),
        different_length_one_empty: (vec![], vec![0.9, 0.1], Err(NegativeLogLikelihoodLossError::InputsHaveDifferentLength)),
        value_greater_than_1: (vec![1.1, 0.0, 0.8], vec![0.1, 0.2, 0.3], Err(NegativeLogLikelihoodLossError::InvalidValues)),
        value_greater_smaller_than_0: (vec![0.9, 0.0, -0.1], vec![0.1, 0.2, 0.3], Err(NegativeLogLikelihoodLossError::InvalidValues)),
        empty_input: (vec![], vec![], Err(NegativeLogLikelihoodLossError::EmptyInputs)),
    }
```
Advantages:
- the code is shorter,
- more _paths_ are tested (6 vs. 8).

Please note, that this change required changing the order of `if`'s.